src/libostree/ostree-repo-prune.c

```diff
@@ -189,6 +189,44 @@ _ostree_repo_prune_tmp (OstreeRepo *self,
   return TRUE;
 }
 
+static gboolean
+get_objects_from_refs_only (OstreeRepo   *repo,
+                            OstreeRepoCommitTraverseFlags flags,
+                            GHashTable   **objects,
+                            GCancellable *cancellable,
+                            GError      **error)
+{
+  *objects =
+  g_hash_table_new_full (ostree_hash_object_name, g_variant_equal,
+                          (GDestroyNotify) g_variant_unref,
+                          (GDestroyNotify) g_variant_unref);
+  g_autoptr(GHashTable) ret_objects = ostree_repo_traverse_new_reachable();
+  g_autoptr(GHashTable) all_refs = NULL;
+  /* We start from the refs */
+  if (!ostree_repo_list_refs (repo, NULL, &all_refs,
+                              cancellable, error))
+    return FALSE;
+  
+  /* Iterate over all refs and add commit objects to 
+   * table of prunable objects.
+   */
+  GLNX_HASH_TABLE_FOREACH (all_refs, const char *, ref)
+    {
+      const char *checksum = g_hash_table_lookup (all_refs, ref);
+      if (!ostree_repo_traverse_commit_with_flags(repo, flags, checksum, -1, ret_objects, NULL, cancellable, error))
+        return FALSE;
+    }
+
+  GLNX_HASH_TABLE_FOREACH (ret_objects, GVariant *, key)
+    {
+      GVariant *value = g_variant_new ("(b@as)", TRUE, g_variant_new_strv (NULL, 0));
+      /* transfer ownership */
+      g_hash_table_replace (*objects, g_variant_ref(key), 
+                            g_variant_ref_sink (value));
+    }
```

Comment:
Can't we directly pass `*objects` to `ostree_repo_traverse_commit_with_flags()`?

Reply:
Oh nevermind, I see...there's some weird historical ostree tech debt here.

Reply:
https://github.com/ostreedev/ostree/pull/2634 - let's rebase this on that PR and do the same thing.